File: backend/app/providers/mock.py
"""Deterministic priced offers. Never used as a substitute for a live shop on other pairs."""

from __future__ import annotations

from datetime import datetime, timezone

from app.models import Offer, Segment
from app.providers.base import ShopRequest


def _seg(
    origin: str,
    dest: str,
    dep: str,
    arr: str,
    flight: str,
    carrier: str,
    minutes: int,
) -> Segment:
    return Segment(
        origin=origin,
        dest=dest,
        carrier=carrier,
        operating_carrier=carrier,
        flight_number=flight,
        dep=dep,
        arr=arr,
        duration_min=minutes,
        rbd="Y",
        aircraft="",
    )


def _offer(
    oid: str,
    price: float,
    segments: list[Segment],
    date: str,
) -> Offer:
    return Offer(
        id=oid,
        kind="nonstop" if len(segments) == 1 else "connecting",
        channel="ndc",
        source="mock",
        layer="priced-offer",
        segments=segments,
        price=price,
        base_price=round(price * 0.82, 2),
        taxes=round(price * 0.18, 2),
        currency="USD",
        cabin="ECONOMY",
        fare_basis="Y",
        seats=7,
        validating_airline=segments[0].carrier,
        refundable=False,
        bags_included=0,
        carrier=segments[0].carrier,
        duration_min=sum(s.duration_min for s in segments) + (40 if len(segments) > 1 else 0),
        stops=max(len(segments) - 1, 0),
        first_flight=segments[0].flight_number,
        retrieved_at=datetime.now(timezone.utc).isoformat(timespec="seconds"),
        note="Mock fixture. Complete priced itinerary — not a reconstructed segment sum.",
    )


def _t(date: str, hhmm: str) -> str:
    return f"{date}T{hhmm}:00"


def jfk_ord_nonstop(date: str) -> Offer:
    return _offer(
        "mock-jfk-ord-nonstop",
        240,
        [_seg("JFK", "ORD", _t(date, "08:00"), _t(date, "10:15"), "AA100", "AA", 135)],
        date,
    )


def jfk_ord_den(date: str) -> Offer:
    return _offer(
        "mock-jfk-ord-den",
        170,
        [
            _seg("JFK", "ORD", _t(date, "08:00"), _t(date, "10:15"), "UA200", "UA", 135),
            _seg("ORD", "DEN", _t(date, "11:30"), _t(date, "13:20"), "UA300", "UA", 110),
        ],
        date,
    )


def jfk_ord_sea(date: str) -> Offer:
    return _offer(
        "mock-jfk-ord-sea",
        185,
        [
            _seg("JFK", "ORD", _t(date, "09:00"), _t(date, "11:20"), "AS400", "AS", 140),
            _seg("ORD", "SEA", _t(date, "12:40"), _t(date, "15:10"), "AS500", "AS", 150),
        ],
        date,
    )


def jfk_dfw_lax(date: str) -> Offer:
    return _offer(
        "mock-jfk-dfw-lax",
        150,
        [
            _seg("JFK", "DFW", _t(date, "07:00"), _t(date, "10:30"), "AA600", "AA", 210),
            _seg("DFW", "LAX", _t(date, "11:40"), _t(date, "13:10"), "AA700", "AA", 90),
        ],
        date,
    )


def jfk_dfw_den(date: str) -> Offer:
    """Refresh mutation: no longer passes through ORD."""
    return _offer(
        "mock-jfk-dfw-den",
        150,
        [
            _seg("JFK", "DFW", _t(date, "08:00"), _t(date, "11:10"), "UA200", "UA", 190),
            _seg("DFW", "DEN", _t(date, "12:20"), _t(date, "13:40"), "UA300", "UA", 80),
        ],
        date,
    )


class MockProvider:
    name = "mock"

    async def shop(self, req: ShopRequest) -> list[Offer]:
        origin, dest = req.origin.upper(), req.dest.upper()
        date = req.date
        if origin == "JFK" and dest in {"ORD", "CHI", "MDW"}:
            return [jfk_ord_nonstop(date), jfk_ord_den(date), jfk_ord_sea(date), jfk_dfw_lax(date)]
        if origin == "JFK" and dest == "DEN":
            return [jfk_ord_den(date)]
        if origin == "JFK" and dest == "SEA":
            return [jfk_ord_sea(date)]
        return []

    async def refresh_offer(self, offer: Offer) -> Offer | None:
        date = (offer.segments[0].dep or "")[:10] or datetime.now(timezone.utc).date().isoformat()
        if offer.id == "mock-jfk-ord-den":
            return jfk_ord_den(date).model_copy(update={"price": 175.0, "id": offer.id})
        if offer.id == "mock-jfk-ord-den-mutated":
            return jfk_dfw_den(date)
        if offer.source != "mock":
            return None
        if offer.price is None:
            return None
        return offer.model_copy(update={"price": round(offer.price + 5, 2)})

File: backend/src/providers/mock.rs
use chrono::Utc;

use crate::models::{Offer, Segment, ShopRequest};

fn seg(origin: &str, dest: &str, dep: &str, arr: &str, flight: &str, carrier: &str, minutes: i32) -> Segment {
    Segment {
        origin: origin.into(),
        dest: dest.into(),
        carrier: carrier.into(),
        operating_carrier: Some(carrier.into()),
        flight_number: flight.into(),
        dep: dep.into(),
        arr: arr.into(),
        duration_min: minutes,
        rbd: "Y".into(),
        aircraft: String::new(),
        fare_basis: None,
    }
}

fn offer(oid: &str, price: f64, segments: Vec<Segment>) -> Offer {
    let n = segments.len();
    let duration = segments.iter().map(|s| s.duration_min).sum::<i32>()
        + if n > 1 { 40 } else { 0 };
    let carrier = segments[0].carrier.clone();
    let first = segments[0].flight_number.clone();
    Offer {
        id: oid.into(),
        kind: if n == 1 {
            "nonstop".into()
        } else {
            "connecting".into()
        },
        channel: "ndc".into(),
        source: "mock".into(),
        layer: "priced-offer".into(),
        segments,
        price: Some(price),
        base_price: Some((price * 0.82 * 100.0).round() / 100.0),
        taxes: Some((price * 0.18 * 100.0).round() / 100.0),
        currency: "USD".into(),
        cabin: "ECONOMY".into(),
        fare_basis: "Y".into(),
        seats: Some(7),
        validating_airline: Some(carrier.clone()),
        carrier,
        duration_min: duration,
        stops: n.saturating_sub(1) as i32,
        first_flight: first,
        retrieved_at: Some(Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Secs, true)),
        note: Some("Mock fixture. Complete priced itinerary — not a reconstructed segment sum.".into()),
        ..Offer::default()
    }
}

fn t(date: &str, hhmm: &str) -> String {
    format!("{date}T{hhmm}:00")
}

pub fn jfk_ord_nonstop(date: &str) -> Offer {
    offer(
        "mock-jfk-ord-nonstop",
        240.0,
        vec![seg("JFK", "ORD", &t(date, "08:00"), &t(date, "10:15"), "AA100", "AA", 135)],
    )
}

pub fn jfk_ord_den(date: &str) -> Offer {
    offer(
        "mock-jfk-ord-den",
        170.0,
        vec![
            seg("JFK", "ORD", &t(date, "08:00"), &t(date, "10:15"), "UA200", "UA", 135),
            seg("ORD", "DEN", &t(date, "11:30"), &t(date, "13:20"), "UA300", "UA", 110),
        ],
    )
}

pub fn jfk_ord_sea(date: &str) -> Offer {
    offer(
        "mock-jfk-ord-sea",
        185.0,
        vec![
            seg("JFK", "ORD", &t(date, "09:00"), &t(date, "11:20"), "AS400", "AS", 140),
            seg("ORD", "SEA", &t(date, "12:40"), &t(date, "15:10"), "AS500", "AS", 150),
        ],
    )
}

pub fn jfk_dfw_lax(date: &str) -> Offer {
    offer(
        "mock-jfk-dfw-lax",
        150.0,
        vec![
            seg("JFK", "DFW", &t(date, "07:00"), &t(date, "10:30"), "AA600", "AA", 210),
            seg("DFW", "LAX", &t(date, "11:40"), &t(date, "13:10"), "AA700", "AA", 90),
        ],
    )
}

pub fn jfk_dfw_den(date: &str) -> Offer {
    offer(
        "mock-jfk-dfw-den",
        150.0,
        vec![
            seg("JFK", "DFW", &t(date, "08:00"), &t(date, "11:10"), "UA200", "UA", 190),
            seg("DFW", "DEN", &t(date, "12:20"), &t(date, "13:40"), "UA300", "UA", 80),
        ],
    )
}

pub fn jfk_ord_roundtrip(date: &str, return_date: &str) -> Offer {
    let mut offer = offer(
        "mock-jfk-ord-roundtrip",
        430.0,
        vec![
            seg("JFK", "ORD", &t(date, "08:00"), &t(date, "10:15"), "AA100", "AA", 135),
            seg("ORD", "JFK", &t(return_date, "18:00"), &t(return_date, "21:15"), "AA101", "AA", 135),
        ],
    );
    offer.kind = "nonstop".into();
    offer.stops = 0;
    offer.return_date = Some(return_date.into());
    offer.outbound_end = Some(0);
    offer.duration_min = 270;
    offer.note = Some("Mock round-trip fixture. Hidden-city is one-way only.".into());
    offer
}

pub struct MockProvider;

impl MockProvider {
    pub fn new() -> Self {
        Self
    }

    pub async fn shop(&self, req: &ShopRequest) -> anyhow::Result<Vec<Offer>> {
        let origin = req.origin.to_uppercase();
        let dest = req.dest.to_uppercase();
        let date = &req.date;
        if req.return_date.is_some() {
            if origin == "JFK" && matches!(dest.as_str(), "ORD" | "CHI" | "MDW") {
                return Ok(vec![jfk_ord_roundtrip(date, req.return_date.as_deref().unwrap())]);
            }
            return Ok(vec![]);
        }
        if origin == "JFK" && matches!(dest.as_str(), "ORD" | "CHI" | "MDW") {
            return Ok(vec![
                jfk_ord_nonstop(date),
                jfk_ord_den(date),
                jfk_ord_sea(date),
                jfk_dfw_lax(date),
            ]);
        }
        if origin == "JFK" && dest == "DEN" {
            return Ok(vec![jfk_ord_den(date)]);
        }
        if origin == "JFK" && dest == "SEA" {
            return Ok(vec![jfk_ord_sea(date)]);
        }
        Ok(vec![])
    }

    pub async fn refresh_offer(&self, offer: &Offer) -> anyhow::Result<Option<Offer>> {
        let date = offer
            .segments
            .first()
            .and_then(|s| {
                if s.dep.len() >= 10 {
                    Some(s.dep[..10].to_string())
                } else {
                    None
                }
            })
            .unwrap_or_else(|| Utc::now().date_naive().to_string());
        if offer.id == "mock-jfk-ord-den" {
            let mut fresh = jfk_ord_den(&date);
            fresh.price = Some(175.0);
            fresh.id = offer.id.clone();
            return Ok(Some(fresh));
        }
        if offer.id == "mock-jfk-ord-den-mutated" {
            return Ok(Some(jfk_dfw_den(&date)));
        }
        if offer.source != "mock" {
            return Ok(None);
        }
        let Some(price) = offer.price else {
            return Ok(None);
        };
        let mut fresh = offer.clone();
        fresh.price = Some(((price + 5.0) * 100.0).round() / 100.0);
        Ok(Some(fresh))
    }
}

impl Default for MockProvider {
    fn default() -> Self {
        Self
    }
}
